### ISSUE TITLE
Hybrid Search (combining Vector + Sparse search)

### ISSUE THREAD

**Issue by laura-ham** [semi-technologies/weaviate#2134]
## WHAT
Combining vector search with sparse (e.g. BM25) search in one query

## WHY
To bridge the gap between sparse search and vector search.

### Longer why
Vector search (using dense vectors, computed by ML models) works well in-domain, but has poor performance out-of-domain.
BM25 search works well out-of-domain, because it uses sparse methods (keyword matching), but can't perform context-based search.
Combining both methods will improve search results out-of-domain.

## HOW
* This issue and implementation depend on issue https://github.com/semi-technologies/weaviate/issues/2133
* Do both a dense and BM25 search using a query (in parallel)
* You should be able to define a function to combine the results into 1 result list, using the scores of data in both candidate lists.
* A default function can be defined in the Weaviate setup, but can be overwritten in the GraphQL query.
* BM25 score is unbounded. Score normalization or scaling is not a good idea, because you lose information on how good the results are textually. However, when combining BM25 with Dense search, some form of normalization may be handy, so we can choose to offer normalization methods anyways. We can explain strategies in the documentation. Possible strategies are:
  * **minmax** - To leave the distribution intact, the best normalization method is a *minmax* approach, which takes the minimum BM25 score and maximum BM25 score into account. Taking the maximum BM25 score of a particular query as maximum in the formula is not a good idea, because then you're setting this result to the maximum score regardless of the actual score. So a (theoretical) maximum needs to be defined, although BM25 is unbounded. This can be achieved by running a number of different queries, and recording the maximum value. Since this is quite complex to implement as a feature, we can start by offering a setting with a default value, which the user can change at runtime. An example can be: `max((x/10), 1)` if the guessed maximum is 10.
  * **Arctangent** - An arctan scales values in a logarithmic manner. A function to scale scores <-1, 1> (practically between <0,1> because x will be a positive score) is: `2/pi*arctan(x)` (see https://www.mdpi.com/2227-7390/10/8/1335)

### Design - Weaviate setup
Requirements:
* Dense retrieval and sparse retrieval independently (with the same or different query)
* Combine the result of both methods using a scoring function

**Schema**
The settings should be configured in the schema, per class:

```json
{
  "class": "string",
  "vectorIndexType": "hnsw",                
  "vectorIndexConfig": {
    ...                                     
  },
  "vectorizer": "text2vec-transformers",
  "moduleConfig": {
    "text2vec-transformers": {  
      "vectorizeClassName": true            
    }
  },                       
  "sparseIndexType": "bm25",                
  "sparseIndexConfig": {                   
    "bm25": {
      "b": 0.75,
      "k1": 1.2
    }
  },
  "properties": [                            
    {
      "name": "string",                     
      "description": "string",              
      "dataType": [                         
        "string"
      ],
      "sparseIndexConfig": {                
        "bm25": {
          "b": 0.75,
          "k1": 1.2, 
        }
      },
      "indexInverted": true                 
    }
  ]
}
```

**Docker-compose**
In case we need to let Weaviate know on startup whether to enable sparse search, we can introduce an env var like `ENABLE_SPARSE_INDEX`: 

```yaml
---
version: '3.4'
services:
  weaviate:
    command:
    - --host
    - 0.0.0.0
    - --port
    - '8080'
    - --scheme
    - http
    image: semitechnologies/weaviate:1.14.1
    ports:
    - 8080:8080
    restart: on-failure:0
    environment:
      QUERY_DEFAULTS_LIMIT: 25
      AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED: 'true'
      PERSISTENCE_DATA_PATH: '/var/lib/weaviate'
      DEFAULT_VECTORIZER_MODULE: text2vec-transformers
      ENABLE_MODULES: text2vec-transformers
      CLUSTER_HOSTNAME: 'node1'
      ENABLE_SPARSE_INDEX: 'true' # <== NEW. Which method (e.g. bm25) can be specified in the schema. Not sure if this variable is needed in the docker-compose actually.
    t2v-transformers:
      image: semitechnologies/transformers-inference:sentence-transformers-msmarco-distilroberta-base-v2
      environment:
        ENABLE_CUDA: 0 # set to 1 to enable
...
```

### Design - GraphQL Queries
The current API is documented in [this comment](https://github.com/semi-technologies/weaviate/issues/2134#issuecomment-1323955096).

Below is the more elaborate original proposal.

```graphql
{
  Get {
    Paper (
      hybridSearch: {               # NEW
        operands: [{
          sparseSearch: {           # inherits all fields from sparseSearch. alternative name: sparse
            function: bm25,
            query: "my query",
            properties: ["abstract"],
            normalization: {
              minmax: {
                max: 10
              }
            }
          },
          weight: 0.2
        }, {
          nearText: {               # inherits all fields from nearText. alternative name: dense or denseSearch
            concepts: ["my query"],
            certainty: 0.7   
          },
          weight: 0.8
        }],
        type: Sum                   # or Average, or RRF (with RRF, weights will be discarded)
      }
    ) {
      title
      abstract
      _additional {
        score    # NEW - because distance is not a good word, it really is a ranking score instead of distance between vectors
      }
    }
  }
}
```

Multiple sparse searches should also be supported, like: 

```graphql
{
  Get {
    Paper (
      hybridSearch: {               # NEW
        operands: [{
          sparseSearch: {           # inherits all fields from sparseSearch. alternative name: sparse
            function: bm25,
            query: "my query",
            properties: ["abstract"],
            normalization: {
              minmax: {
                max: 10
              }
            }
          },
          weight: 0.2
        }, {
          sparseSearch: {           # inherits all fields from sparseSearch. alternative name: sparse
            function: bm25,
            query: "my query 2",
            properties: ["title"],
            normalization: {
              minmax: {
                max: 10
              }
            },
            limit: 100
          },
          weight: 0.2
        }, {
          nearText: {              # inherits all fields from nearText. alternative name: dense or denseSearch
            concepts: ["my query"],
            certainty: 0.7,
            limit: 100   
          },
          weight: 0.6
        }],
        type: Sum                  # or Average, or RRF (with RRF, weights will be discarded)
      }
    ) {
      title
      abstract
      _additional {
        score        # NEW - because distance is not a good word, it really is a ranking score instead of distance between vectors
      }
    }
  }
}
```

**Comment by bobvanluijt** [comment]
Thanks @laura-ham – some thoughts

1. Is `ENABLE_SPARSE_INDEX` needed? Or can we enable it by default (like one doesn't have to enable the inverted index)
2. I like the idea of inherence. Question: _does GraphQL know the order of an array on input?_
3. I'm not 100% sure I understand the "inherence flow" can you give an example of this?

**Comment by laura-ham** [comment]
> 1. Is ENABLE_SPARSE_INDEX needed? Or can we enable it by default (like one doesn't have to enable the inverted index)

Yes I thought the same (hence my comment in that line). Curious to @etiennedi's opinion here implementation wise ;) 

> 2. I like the idea of inherence. Question: does GraphQL know the order of an array on input?

This design will not look at the order, but just does both searches independently of each other, and combines the results later. The order of different search stages will follow in another proposal of introducing search pipelines. 

> 3. I'm not 100% sure I understand the "inherence flow" can you give an example of this?

What do you mean with "inherence flow? In the first example I gave it will be like: 
1. Do a sparse and vector retrieval and get (+normalize) the scores
2. Combine the results with the linear sum formula (score = 0.2 * sparse_results + 0.8 * vector_results).
3. The results will be presented to the user in descending order of the final score

**Comment by bobvanluijt** [comment]
My "inherence flow" question is answered by: "this design will not look at the order"

For all other answers: _[thanks](https://www.pinterest.com/pin/862087553649959763/)_

**Comment by etiennedi** [comment]
Hi, a couple of questions/thoughts from my side:

* I believe we don't need the `ENABLE_SPARSE_INDEX`, what to index can already be controlled in the schema. By default, any text field would be indexed for sparse search.
* Does the current calculation produce good results? I find it very hard to form an opinion on whether the API is appropriate or not, without knowing if we can achieve good results this way? Did you already run some experiments of results that you can share
* Agree with `score` being better than distance, the moment that combining/ranking/boosting/etc is involved.
* How do we currently get from distance to score for the dense part? Would it just be `score = - distance`?

**Comment by etiennedi** [comment]
Something that popped into my head: Should there be a field that controls the definition of "hybrid". For example, right now, "hybrid" is defined as "two independent searches with scores merged". But I could imagine that in the future we might also want to:

* Let BM25 act as a filter (e.g. the top n results are the candidate set), then perform a filtered vector search
* Perform a vector search, and apply some sort of a limit, then boost results based on BM25 scores (cc @CShorten who has mentioned boosting before). Not sure yet how feasible this is from a tech perspective, but it sounds like it could come in handy in the future.

Assuming that we won't have a complex pipelining setup, covering all three with the same API might lead to an easy-to-use UX that can be quite powerful.

**Comment by bobvanluijt** [comment]
Some feedback based on experiments with the 1st test version and 

### Default `alpha` setting

I understand that the value should be different per use case, but a default setting for alpha of `0.25` seems appropriate.

### Adding `query` to `nearText`

To stay consistent, adding `query` to `nearText` seem appropriate.

E.g.,

```
{
  Get {
    PageChunk(
      nearText: {
        query: "how to import data?"
```

### Align syntax

We now have:

* `hybridSearch`
* `bm25`
* `nearText`

That's not very consistent. I like the word `Search` in `hybridSearch` but it seems slightly pleonastic. Because this would result in: `hybridSearch`, `bm25Search`, etc.

I'm proposing:

* `hybridSearch` => rename to: `hybrid`
* `bm25` => keep as is
* `nearText` => keep as is

### Allow for bring-your-own-vector

I think that an example shows this best:

```
{
  Get {
    PageChunk(
      hybridSearch: {
        vector: [0.1, 0.2]
        alpha: 0.25
```

### BM25 doesn't do exact matches?

This results in a bunch of exact matches:

```
{
  Get {
    PageChunk(
      nearText: {
        concepts: ["text2vec-openai"]
      }
```

However, this result in `0` results:

```
{
  Get {
    PageChunk(
      bm25: {
        query: "text2vec-transformers"
        properties: ["content"]
      }
```

I'm confused about this.

HybridSearch works well.

**Comment by etiennedi** [comment]
Some comments for the other points:

> To stay consistent, adding query to nearText seem appropriate.

I really like this. `nearText.concepts` isn't nearly as intuitive as `nearText.query` imho. This does, however, raise the question of what to do with the existing concepts. I see the following options:

1. Keep both, in docs keep preferring `concepts`
2. Keep both, in docs prefer `query`
3. Keep both, in docs always show both as equal alternatives
4. Keep both, but officially consider `concepts` deprecated, in docs exclusively use `query`

Out of those I have a preference for 2 or 4. I don't like 3. I'm not sure if deprecating `concepts` is necessary, because `concepts` takes an array, whereas `query` takes a single string. So they're not identical and an argument could be made that both should exist. So option 2?

> Align syntax

Agreed, I like it better without `Search` everywhere.

> Allow for bring-your-own-vector

Check, this is the same syntax that we also proposed in Italy.

> BM25 doesn't do exact matches?
> 
> I'm confused about this.
> 
> HybridSearch works well.

No idea what's going on here – will investigate. What's needed to reproduce this?

**Comment by bobvanluijt** [comment]
> So with alpha==0.25 are you saying by default it should weight sparse more heavily than dense? I would have expected it the other way round?

You are correct, this is my mistake 🙈 

> To stay consistent, adding query to nearText seem appropriate.

Option 2 (please align with @sebawita on this as well)

> _the rest_

👍

**Comment by sebawita** [comment]
are we sticking with `alpha`?
I am not sure if that attribute is self-explanatory... the fact that we are not sure what 0.25 should mean, already tells us that perhaps the name is not intuitive

Also, is this always going to be a two-way split? or do we expect scenarios with 3-way (or more) queries?
Because in this case, `alpha` won't cover the ratio split.

**Comment by etiennedi** [comment]
For completeness' sake, and someone outside of SeMI viewing this: This is the most up-to-date proposal for the current API (as opposed to the more elaborate proposal at the beginning of this issue). This should already include the suggestions @bobvanluijt made [here](https://github.com/semi-technologies/weaviate/issues/2134#issuecomment-1320159135).

Whether or not the ratio parameter should be called `alpha` or something else is currently up for debate (see posts before this). 

```graphql
{
  Get {
    <ClassName>(
      hybrid: {
        query: "my text query"
        vector: [0.1, 0.2] # optional for BYOV cases
        alpha: 0.5 # optional – defaults to 0.75
      }
    ){
       prop_1
       prop_2   
    }
  }
}
```

**Comment by etiennedi** [comment]
@bobvanluijt 

Regarding

> BM25 doesn't do exact matches?
> I'm confused about this.
> HybridSearch works well.

I can confirm this issue. This is because the pre-processing and tokenization rules at query time and index time are not in sync at the moment. There is an issue on the [internal issue tracker for this here](https://semi-technology.atlassian.net/browse/WEAVIATE-382?atlOrigin=eyJpIjoiNjJlMDhhMzAwZTI3NDViNWJiMWE2ZTY3ZTg1ZWFkMDYiLCJwIjoiamlyYS1zbGFjay1pbnQifQ). This will, of course, be fixed before the public release.

**Comment by donomii** [comment]
> Some feedback based on experiments with the 1st test version and

> ### Adding `query` to `nearText`
> To stay consistent, adding `query` to `nearText` seem appropriate.
> 
> E.g.,
> 
> ```
> {
>   Get {
>     PageChunk(
>       nearText: {
>         query: "how to import data?"
> ```

This is great, but it should probably go in its own ticket

> 
> I'm proposing:
> 
> * `hybridSearch` => rename to: `hybrid`
> * `bm25` => keep as is
> * `nearText` => keep as is
> 
> ### Allow for bring-your-own-vector
> I think that an example shows this best:
> 
> ```
> {
>   Get {
>     PageChunk(
>       hybridSearch: {
>         vector: [0.1, 0.2]
>         alpha: 0.25
> ```

There is now a vector option that does exactly this.

Limits should also work now

**Comment by zoltan-fedor** [comment]
Hi @etiennedi and @bobvanluijt ,
I am very happy to see that the hybrid search is progressing.

I wanted to highlight one issue with the hybrid search - namely that for it to work, the BM25 (sparse) and the vector search should support the same capabilities.

For example I believe currently the BM25 search does NOT support filtering (eg BM25 search being combined with a filter), see issue I have raised in July: https://github.com/semi-technologies/weaviate/issues/2061

This would mean, that the hybrid search couldn't support filtering either, as you cannot have only one side of the hybrid search (the vector search) do the filtering, while the BM25 search to bring in unfiltered results, as combining those two results would mean that records which don't meet the criteria set in the filter would make it into the hybrid search's results.

**Comment by etiennedi** [comment]
Hi @zoltan-fedor, thanks, and this is great feedback! In v1.17 BM25 and dense will not have feature parity yet. Filters are indeed still missing from BM25 (and therefore currently also from the Hybrid API). But we will certainly add them in one of the next relases. 

I'll create an issue so you can upvote it, one sec.

**Comment by etiennedi** [comment]
You can upvote it [here](https://github.com/semi-technologies/weaviate/issues/2393).

**Comment by zoltan-fedor** [comment]
Thanks @etiennedi !
I think this (filters) is important to have, as I see that one of the main selling points of using Weaviate - allowing you to combine vector (and sparse) search with filtering.

**Comment by etiennedi** [comment]
Yes, absolutely agreed. We don't want to wait for every possible feature to be in before releasing. Rather we'll have a few releases that build the feature up more and more. In v1.17 (three weeks from now) we'll have the initial (minimal) version. It can already cover a lot of cases and we'll get early feedback. And then, based on user feedback, such as yours, we'll add more capabilities, including filtering for BM25.

**Comment by zoltan-fedor** [comment]
I get it, I just thought that the BM25 being in experimental mode since Q2 was the period during which you were expecting user feedback (which is why I raised this in July https://github.com/semi-technologies/weaviate/issues/2061) and releasing it into production (https://github.com/semi-technologies/weaviate/issues/2225) is by when you would want all the basic compatibility functionalities covered.

No worries, I guess we just have a slightly different definitions of the purpose of experimental mode vs production-readyness in our projects ... :-)

Either way, I am very much looking forward to have the feature-parity gap closed between the vector search and BM25&Hybrid after v1.17 then!

Ps.: Even the Haystack-Weaviate integration module of Haystack where I have added the BM25 search back in July have this shortcoming noted (no filtering with BM25 on the Weaviate document store), so once that gap is closed, I will need to remove that from the Haystack codebase. I was hoping to remove that soone, as I have seen that a lot of new Haystack users being initially interested using Weaviate as the document store, but this was a limiting factor in their ability to pick Weaviate and so usually ended up going with something else.

**Comment by etiennedi** [comment]
You do raise a very good point and even if it won't be available in the next release just yet - the priority this has for you (and the haystack users you mentioned) is duly noted and the reminder very much appreciated. 

PS: Thanks for linking your original issue - I had forgotten it existed, I'll close mine again.

**Comment by zoltan-fedor** [comment]
Hah, I just linked from my old issue to your newly created one, as I thought maybe the one you created will get more attention than mine.  Hah :-)

**Comment by etiennedi** [comment]
Haha, Race condition. I saw your comment and did indeed close yours as you had already upvoted mine ;-)

**Comment by zoltan-fedor** [comment]
@etiennedi ,
I am getting confused, is the filtering with BM25 is already supported?
I was just looking at my Haystack code and saw that just a few days ago it was modified to remove the limitation (throwing an error when somebody tries to use filters with BM25 on Weaviate):
https://github.com/deepset-ai/haystack/pull/3628

**Comment by etiennedi** [comment]
Don't know what the motivation behind the linked PR is, but I just validated with the latest build from the hybrid branch and it errors the same as it did before. I knew we haven't built it yet, but I just wanted to double-check in case I had forgotten ;-) But it does show the "correct" error message.

<img width="1768" alt="image" src="https://user-images.githubusercontent.com/8974479/204638022-a3b1163c-0518-4092-870b-82f590c9bd8d.png">

**Comment by zoltan-fedor** [comment]
Okey, then that Haystack PR cannot be correct.
That is odd, as it has a unittest which I wrote back in July which catches the error thrown by Weaviate and they have uncommented the exception capture in that PR.
https://github.com/deepset-ai/haystack/pull/3628/files#diff-bc2015fb1dc5e70c32cf9c8bba6a2497388e0293d385cb4e614cd9cea534b679L158

Something doesn't square.

**Comment by zoltan-fedor** [comment]
Hi @etiennedi ,
As suspected, that Haystack PR was wrong, incorrectly assumed that Weaviate now supports filters with BM25 (and also included a bug causing it in reality run an unfiltered BM25 search).
I have asked them to remove that PR, as that would introduce a bug into Haystack.
See my rather lengthy explanation at https://github.com/deepset-ai/haystack/pull/3628#issuecomment-1331269953

On the side note, this just underlines my earlier statement, that this is an important functionality, that is really needed in Weaviate and we certainly aren't the only ones missing it...
